### PR TITLE
delete file by mistake when rewriting iceberg table  throw CommitStateUnknownException

### PR DESCRIPTION
**Pull request by wg1026688210**
This pr is to solve the problem of deleting files by mistake in rewriting table  if we can’t judge whether the submission is successful when commiting. 
The solution is that we only print some logs and do not  clean up the files  if throw CommitStateUnknownException when commiting ,leave the cleanup work to RemoveOrphanFiles

### CODE REVIEW

File: core/src/main/java/org/apache/iceberg/actions/BaseRewriteDataFilesAction.java [hunk 1]
```diff
@@ -271,20 +272,30 @@ public RewriteDataFilesActionResult execute() {
   private void replaceDataFiles(Iterable<DataFile> deletedDataFiles, Iterable<DataFile> addedDataFiles,
                                 long startingSnapshotId) {
     try {
-      RewriteFiles rewriteFiles = table.newRewrite()
-          .validateFromSnapshot(startingSnapshotId)
-          .rewriteFiles(Sets.newHashSet(deletedDataFiles), Sets.newHashSet(addedDataFiles));
-      commit(rewriteFiles);
+      doReplace(deletedDataFiles, addedDataFiles, startingSnapshotId);
```
Comment (stevenzwu) [inline core/src/main/java/org/apache/iceberg/actions/BaseRewriteDataFilesAction.java:275]: nit: wondering if it is necessary to extract this into a `doReplace` method. the current method is not long at all and doReplace is only used once here.

**Review by stevenzwu — APPROVED** [review]
left one minor comment

File: core/src/main/java/org/apache/iceberg/actions/BaseRewriteDataFilesAction.java (hunk 1 above)
Comment (wg1026688210) [reply core/src/main/java/org/apache/iceberg/actions/BaseRewriteDataFilesAction.java:275]: i'm sorry for no setting  the pr as a draft, the method is for a  unit test

**Comment by wg1026688210** [comment]
cc @stevenzwu @openinx @rdblue 

> left one minor comment

thanks for your concern, it's ready for review

File: core/src/main/java/org/apache/iceberg/actions/BaseRewriteDataFilesAction.java [hunk 2]
```diff
@@ -252,7 +254,6 @@ public RewriteDataFilesActionResult execute() {
     return new RewriteDataFilesActionResult(currentDataFiles, addedDataFiles);
   }
 
-
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/actions/BaseRewriteDataFilesAction.java:255]: Please remove unnecessary changes like this and reformatting lines that don't need to change. We avoid changes that aren't required to avoid git commit conflicts.

**Comment by rdblue** [comment]
This looks valid to me. The action will clean up data files for any exception, but we should not do that when the commit state is unknown. FYI @RussellSpitzer.

@wg1026688210, can you remove any unnecessary formatting and whitespace changes? It would also be great to update the description with the case that this fixes. Thanks for working on this!

**Comment by RussellSpitzer** [comment]
Yeah this is definitely the right thing to do, we are handling this in the new code (https://github.com/apache/iceberg/blob/a1bd63d56751999ecee89b871992d7bac395fd52/core/src/main/java/org/apache/iceberg/actions/RewriteDataFilesCommitManager.java#L97-L109) as well and while this path is deprecated now I think we should definitely fix it to avoid causing issues while the path is still available.

File: core/src/main/java/org/apache/iceberg/actions/BaseRewriteDataFilesAction.java (hunk 1 above)
Comment (RussellSpitzer) [reply core/src/main/java/org/apache/iceberg/actions/BaseRewriteDataFilesAction.java:276]: I'm not sure if the modules work out correctly for this, but it may make sense to use RewriteDataFilesCommitter (https://github.com/apache/iceberg/blob/a1bd63d56751999ecee89b871992d7bac395fd52/core/src/main/java/org/apache/iceberg/actions/RewriteDataFilesCommitManager.java#L57) here and re-use the commit code from the new path.

Comment (RussellSpitzer) [reply core/src/main/java/org/apache/iceberg/actions/BaseRewriteDataFilesAction.java:276]: Actually this may be more complicated than just doing the fix the way you have already done it here ... so maybe forget that

File: core/src/main/java/org/apache/iceberg/actions/BaseRewriteDataFilesAction.java (hunk 2 above)
Comment (wg1026688210) [reply core/src/main/java/org/apache/iceberg/actions/BaseRewriteDataFilesAction.java:255]: > Yeah this is definitely the right thing to do, we are handling this in the new code (
> 
> https://github.com/apache/iceberg/blob/a1bd63d56751999ecee89b871992d7bac395fd52/core/src/main/java/org/apache/iceberg/actions/RewriteDataFilesCommitManager.java#L97-L109
> 
> ) as well and while this path is deprecated now I think we should definitely fix it to avoid causing issues while the path is still available.

looking forward to using the new version rewrite

**Comment by wg1026688210** [comment]
> This looks valid to me. The action will clean up data files for any exception, but we should not do that when the commit state is unknown. FYI @RussellSpitzer.
> 
> @wg1026688210, can you remove any unnecessary formatting and whitespace changes? It would also be great to update the description with the case that this fixes. Thanks for working on this!

ok

**Comment by wg1026688210** [comment]
> Yeah this is definitely the right thing to do, we are handling this in the new code (
> 
> https://github.com/apache/iceberg/blob/a1bd63d56751999ecee89b871992d7bac395fd52/core/src/main/java/org/apache/iceberg/actions/RewriteDataFilesCommitManager.java#L97-L109
> 
> ) as well and while this path is deprecated now I think we should definitely fix it to avoid causing issues while the path is still available.

looking forward to using the new version rewrite

File: core/src/main/java/org/apache/iceberg/actions/BaseRewriteDataFilesAction.java [hunk 3]
```diff
@@ -268,23 +270,36 @@ public RewriteDataFilesActionResult execute() {
     return tasksGroupedByPartition.asMap();
   }
 
-  private void replaceDataFiles(Iterable<DataFile> deletedDataFiles, Iterable<DataFile> addedDataFiles,
-                                long startingSnapshotId) {
+  private void replaceDataFiles(
+      Iterable<DataFile> deletedDataFiles, Iterable<DataFile> addedDataFiles,
+      long startingSnapshotId) {
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/actions/BaseRewriteDataFilesAction.java:275]: @wg1026688210, please remove changes like this that are unnecessary.

File: core/src/main/java/org/apache/iceberg/actions/BaseRewriteDataFilesAction.java [hunk 4]
```diff
@@ -268,23 +270,36 @@ public RewriteDataFilesActionResult execute() {
...
-      commit(rewriteFiles);
+      doReplace(deletedDataFiles, addedDataFiles, startingSnapshotId);
     } catch (Exception e) {
-      Tasks.foreach(Iterables.transform(addedDataFiles, f -> f.path().toString()))
-          .noRetry()
-          .suppressFailureWhenFinished()
-          .onFailure((location, exc) -> LOG.warn("Failed to delete: {}", location, exc))
-          .run(fileIO::deleteFile);
+      if (e instanceof CommitStateUnknownException) {
+        LOG.warn("for unknown commiting state ,we should not delete file ");
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/actions/BaseRewriteDataFilesAction.java:280]: Please copy the other warning messages here. This needs to conform more to the error message expectations that we use:
* Messages should use sentence case
* Do not use personal pronouns
* Be clear about what happened and what action is being taken

This should be something more like "Skipping commit cleanup: commit state is unknown". But check the other places where we catch `CommitStateUnknownException` and use a similar message.

File: core/src/main/java/org/apache/iceberg/actions/BaseRewriteDataFilesAction.java [hunk 5]
```diff
@@ -268,23 +270,36 @@ public RewriteDataFilesActionResult execute() {
...
-          .rewriteFiles(Sets.newHashSet(deletedDataFiles), Sets.newHashSet(addedDataFiles));
-      commit(rewriteFiles);
+      doReplace(deletedDataFiles, addedDataFiles, startingSnapshotId);
     } catch (Exception e) {
-      Tasks.foreach(Iterables.transform(addedDataFiles, f -> f.path().toString()))
-          .noRetry()
-          .suppressFailureWhenFinished()
-          .onFailure((location, exc) -> LOG.warn("Failed to delete: {}", location, exc))
-          .run(fileIO::deleteFile);
+      if (e instanceof CommitStateUnknownException) {
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/actions/BaseRewriteDataFilesAction.java:279]: Why is this not a separate `catch` block? I think it should be a `catch` instead of using `instanceof`.

File: core/src/main/java/org/apache/iceberg/actions/BaseRewriteDataFilesAction.java (hunk 4 above)
Comment (wg1026688210) [reply core/src/main/java/org/apache/iceberg/actions/BaseRewriteDataFilesAction.java:280]: thanks, learned a lot for me

File: core/src/main/java/org/apache/iceberg/actions/BaseRewriteDataFilesAction.java [hunk 6]
```diff
@@ -271,11 +273,13 @@ public RewriteDataFilesActionResult execute() {
...
                                 long startingSnapshotId) {
     try {
-      RewriteFiles rewriteFiles = table.newRewrite()
-          .validateFromSnapshot(startingSnapshotId)
-          .rewriteFiles(Sets.newHashSet(deletedDataFiles), Sets.newHashSet(addedDataFiles));
-      commit(rewriteFiles);
+      doReplace(deletedDataFiles, addedDataFiles, startingSnapshotId);
+    } catch (CommitStateUnknownException e) {
+      LOG.warn("Commit state unknown for files: {}, cannot clean up files because they may have been committed " +
+          "successfully.", Lists.newArrayList(addedDataFiles), e);
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/actions/BaseRewriteDataFilesAction.java:279]: Let's remove the list of data files from logs. That is going to be too long. Instead, let's use "Commit state unknown, cannot clean up files that may have been committed"

File: core/src/main/java/org/apache/iceberg/actions/BaseRewriteDataFilesAction.java [hunk 7]
```diff
@@ -271,11 +273,13 @@ public RewriteDataFilesActionResult execute() {
...
+      throw e;
     } catch (Exception e) {
+      LOG.warn("Failed to rewrite, files {} will be delete", Lists.newArrayList(addedDataFiles));
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/actions/BaseRewriteDataFilesAction.java:282]: Let's remove the file list here as well. Instead, use "Failed to commit rewrite, cleaning up rewritten files"

**Review by rdblue — APPROVED** [review]

**Comment by wg1026688210** [comment]
Hi,@rdblue @RussellSpitzer .the pr appears conflicts. Do we need to maintain the  rewrite api of  old  version  next release.  If necessary, I will resolve this conflict.

**Comment by rdblue** [comment]
I'll defer to @RussellSpitzer on whether we should fix the old API.

**Comment by RussellSpitzer** [comment]
Sorry for taking so long to get to this, I think this would be worth keeping just because of how difficult it is to deal with this situation. Please update the PR and I will merge it in.

**Comment by wg1026688210** [comment]
> 
Ok, I will resolve this conflict today

**Comment by wg1026688210** [comment]
> Sorry for taking so long to get to this, I think this would be worth keeping just because of how difficult it is to deal with this situation. Please update the PR and I will merge it in.

hi~ @RussellSpitzer the  conflict has been resolved

**Review by RussellSpitzer — APPROVED** [review]
Thanks for the fix!
